File: app/src/main/java/com/maximillionsnyder/umafinidad/data/Dtos.kt
package com.maximillionsnyder.umafinidad.data

import com.maximillionsnyder.umafinidad.domain.Character
import com.maximillionsnyder.umafinidad.domain.Member
import com.maximillionsnyder.umafinidad.domain.Relation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
internal data class CharacterDto(
    @SerialName("char_id") val charId: Int,
    @SerialName("en_name") val enName: String? = null,
    @SerialName("jp_name") val jpName: String? = null,
    val playable: Boolean? = null,
    val active: Boolean? = null,
)

@Serializable
internal data class RelationDto(
    @SerialName("relation_type") val relationType: Int,
    @SerialName("relation_point") val relationPoint: Int,
)

@Serializable
internal data class MemberDto(
    @SerialName("chara_id") val charaId: Int,
    @SerialName("relation_type") val relationType: Int,
)

@Serializable
internal data class DataBundle(
    val characters: List<CharacterDto>,
    val relations: List<RelationDto>,
    val members: List<MemberDto>,
)

/* Configuración del parser: los JSON datamined traen muchísimos campos que
   la app no usa. */
val jsonParser: Json = Json { ignoreUnknownKeys = true }

internal fun CharacterDto.toDomain() = Character(charId, enName, jpName, playable, active)
internal fun RelationDto.toDomain() = Relation(relationType, relationPoint)
internal fun MemberDto.toDomain() = Member(charaId, relationType)

File: sitio/src/lib/data/dtos.ts
/* DTOs del JSON datamined (snake_case) y conversión al dominio camelCase.
   Espejo de data/Dtos.kt. El parser ignora los campos que no se usan. */

import type { Character, Member, Relation } from '../domain/models'

interface CharacterDto {
  char_id: number
  en_name?: string | null
  jp_name?: string | null
  playable?: boolean | null
  active?: boolean | null
  url_name?: string | null
}

interface RelationDto {
  relation_type: number
  relation_point: number
}

interface MemberDto {
  chara_id: number
  relation_type: number
}

export function parseCharacters(raw: unknown): Character[] {
  const lista = raw as CharacterDto[]
  return lista.map((d) => ({
    charId: d.char_id,
    enName: d.en_name ?? null,
    jpName: d.jp_name ?? null,
    playable: d.playable ?? null,
    active: d.active ?? null,
    urlName: d.url_name ?? null,
  }))
}

export function parseRelations(raw: unknown): Relation[] {
  return (raw as RelationDto[]).map((d) => ({
    relationType: d.relation_type,
    relationPoint: d.relation_point,
  }))
}

export function parseMembers(raw: unknown): Member[] {
  return (raw as MemberDto[]).map((d) => ({
    charaId: d.chara_id,
    relationType: d.relation_type,
  }))
}

/* aptitudes.json: mapa char_id (como string) → 10 letras A–G. Las claves
   se convierten a Int para alinearlas con el resto del modelo. */
export function parseAptitudes(raw: unknown): Map<number, string[]> {
  const mapa = raw as Record<string, string[]>
  return new Map(Object.entries(mapa).map(([id, apt]) => [Number(id), apt]))
}
